File: src/main/java/scampi/cp/constraints/ElementVar.java
/*******************************************************************************
 * This program and the accompanying materials
 * are made available under the terms of the GNU Lesser Public License v3
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/lgpl.html
 *  
 * Contributors:
 *      www.n-side.com
 ******************************************************************************/
package scampi.cp.constraints;

import scampi.cp.core.CPOutcome;
import scampi.cp.core.CPPropagStrength;
import scampi.cp.core.Constraint;
import scampi.cp.core.CPVarInt;

/**
 * @author Pierre Schaus pschaus@gmail.com
 */
public class ElementVar extends Constraint {
	
	CPVarInt [] y;
	CPVarInt x;
	CPVarInt z;

	//y[x] = z
	public ElementVar(CPVarInt [] y, CPVarInt x, CPVarInt z) {
		super(y[0].getStore(),"ElementVar");
		this.y = y;
		this.x = x;
		this.z = z;
	}
	
	//y[x] = z
	public ElementVar(CPVarInt [] y, CPVarInt x, int z) {
		this(y,x,new CPVarInt(x.getStore(), z, z));
	}

	@Override
	protected CPOutcome setup(CPPropagStrength l) {
		if (x.updateMin(0) == CPOutcome.Failure) {
			return CPOutcome.Failure;
		}
		if (x.updateMax(y.length-1) == CPOutcome.Failure) {
			return CPOutcome.Failure;
		}
		if (propagate() == CPOutcome.Failure) {
		      return CPOutcome.Failure;
		}
		z.callPropagateWhenBoundsChange(this);
		x.callPropagateWhenDomainChanges(this);			
		for (Integer v: x) {
			y[v].callPropagateWhenBoundsChange(this);
		}
		return CPOutcome.Suspend;
	}
	
	@Override
	protected CPOutcome propagate() {
		   // z = y[x] 

		   int minY = Integer.MAX_VALUE;
		   int maxY = Integer.MIN_VALUE;
		   for (Integer v: x) {
			   minY = Math.min(minY, y[v].getMin());
			   maxY = Math.max(maxY, y[v].getMax());
		   }

		   // update z
		   if (z.updateMin(minY) == CPOutcome.Failure)
		      return CPOutcome.Failure;
		   if (z.updateMax(maxY) == CPOutcome.Failure)
		      return CPOutcome.Failure;
		   
		   // update x
		   for (Integer v: x) {
			   if (y[v].getMin() > z.getMax() || y[v].getMax() < z.getMin()) {
				   if (x.removeValue(v) == CPOutcome.Failure) {
					   return CPOutcome.Failure;
				   }
			   }
		   }
		   
		   if (x.isBound()) { // x is bound
		      CPVarInt yx = y[x.getValue()];
		      if (yx.updateMin(z.getMin()) == CPOutcome.Failure)
		         return CPOutcome.Failure;
		      if (yx.updateMax(z.getMax()) == CPOutcome.Failure)
		         return CPOutcome.Failure;
		   }
		   return CPOutcome.Suspend;
	}

}

File: oscar-cp/src/main/scala/oscar/cp/constraints/ElementVar.scala
/*******************************************************************************
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *   
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License  for more details.
 *   
 * You should have received a copy of the GNU Lesser General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
 ******************************************************************************/


package oscar.cp.constraints;

import oscar.cp.core.CPOutcome
import oscar.cp.core.CPPropagStrength
import oscar.cp.core.CPIntVar
import oscar.cp.core.Constraint
import oscar.cp.util.ArrayUtils;
import oscar.algo.reversible.ReversibleInt

import scala.math.min
import scala.math.max

import oscar.cp.core._
import oscar.cp.core.CPOutcome._
import oscar.cp.modeling.CPSolver
import oscar.algo.reversible.ReversibleInt
import oscar.algo.reversible.ReversibleSetIndexedArray


/**
 * Element Constraint: y(x) == z
 * Use CPPropagStrength.Strong to have GAC propagation, otherwise BC propagation is used.
 * @author Pierre Schaus - pschaus@gmail.com
 */
class ElementVar(val y: Array[CPIntVar], val x: CPIntVar, val z: CPIntVar) extends Constraint(y(0).s, "ElementVar") {
    


  override def setup(l: CPPropagStrength): CPOutcome = {
    if (l == CPPropagStrength.Strong) {
      if (s.post(new ElementVarAC(y,x,z)) == Failure) return Failure
      else Success
    } else {
      if (s.post(new ElementVarBC(y,x,z)) == Failure) return Failure
      else Success      
    }

  }
  
 

}
